src/lib.rs

```diff
@@ -1,7 +1,552 @@
+#![allow(non_snake_case)]
+
+mod ed25519;
+mod secp256k1;
+
+use crate::{
+    ed25519::{H, H_PRIME},
+    secp256k1::{
+        g,
+        marker::{Mark, Secret, Zero},
+        G, G_PRIME,
+    },
+};
+use bit_vec::BitVec;
+use ecdsa_fun::fun::{
+    marker::{Jacobian, NonZero, Normal},
+    s,
+};
+use rand::{CryptoRng, RngCore};
+use sha2::{Digest, Sha256};
+use std::ops::{Add, Sub};
+
+/// A scalar that is valid for both secp256k1 and ed25519.
+///
+/// Any valid scalar for ed25519 has the same bit representation for secp256k1,
+/// due to the smaller curve order for ed25519 compared to secp256k1.
+///
+/// On the other hand, not all valid scalars for secp256k1 have the same bit
+/// representation for ed25519.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Scalar([u8; 32]);
+
+impl Scalar {
+    /// Generate a random scalar.
+    ///
+    /// To ensure that the scalar is valid and equal for both secp256k1 and
+    /// ed25519, we delegate to an `ed25519::Scalar` API.
+    pub fn random<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
+        let ed25519 = ed25519::Scalar::random(rng);
+        let bytes = ed25519.to_bytes();
+
+        Self(bytes)
+    }
+
+    /// Decompose scalar into bits.
+    ///
+    /// The vector of bits is ordered from least significant bit to most
+    /// significant bit.
+    pub fn bits(&self) -> BitVec {
+        // We reverse the vector of bits to ensure that the bits are
+        // ordered from LSB to MSB.
+        BitVec::from_bytes(&self.0).iter().rev().collect()
+    }
+
+    pub fn into_secp256k1(self) -> secp256k1::Scalar {
+        self.into()
+    }
+
+    pub fn into_ed25519(self) -> ed25519::Scalar {
+        self.into()
+    }
+}
+
+/// Implement addition for `Scalar`.
+///
+/// We choose to rely on `secp256k1::Scalar` for this, but it should be
+/// equivalent to use `ed25519::Scalar` instead.
+impl Add<Scalar> for Scalar {
+    type Output = Scalar;
+    fn add(self, rhs: Scalar) -> Self::Output {
+        let res = s!({ self.into_secp256k1() } + { rhs.into_secp256k1() });
```

Comment:
For OR statement  you actually want to do the addition modulo the smaller curve's modulus (to avoid the prover being able to choose multiple values for its _real_ challenge but have the _fake_ challenge reduce to the same thing regardless modulo ed25519). Or better yet, just use a 31 byte challenge and do bitwise xor as the operation in the constriant see Shoenmakers 5.2.4.

Reply:
> For OR statement you actually want to do the addition modulo the smaller curve's modulus (to avoid the prover being able to choose multiple values for its real challenge but have the fake challenge reduce to the same thing regardless modulo ed25519).

Very true.

I actually think what I'm doing here is just wrong (@tcharding pointed out something along these lines too).

The "promise" of this scalar type is that it always remains modulo ed25519. To uphold that promise I must always operate modulo ed25519.

Thanks for highlighting the problem!

> Or better yet, just use a 31 byte challenge and do bitwise xor as the operation in the constriant see Shoenmakers 5.2.4.

Why do you think that would be better? And why would the challenge need to be 31 bytes in that case?

Reply:
The relationship between the real and fake challenges has nothing to do with the curves other than the resulting challenges should never be greater than either curve's modulus (or if it is -- it should be much greater so not to create a bias after it is reduced). Having 31 bytes just ensures no modular reduction applies before creating the response. XOR is just a faster to compute and simpler linear constraint on the challenges (rather than addition modulo whatever).

Reply:
That makes a lot of sense.